### PR TITLE
BE & FE | Events Listing Page

### PR DESCRIPTION
**Pull request by SharmaineLim**
[Link to Ticket](https://torchbox.monday.com/boards/1192293412/pulses/1400769730)

### Description of Changes Made

- Style the Events Listing Page according to the [design](https://www.figma.com/file/QdMXBHr57mN1svsaNQ7Ggj/Torchbox-2024-(Master)?type=design&mode=design&t=HylHeCsXH62Q7Vd2-0).
- Tweak the model to use `EventItemBlock`, which is based on the same event block used in service pages.

Related changes:

- Create `BaseEventBlock`, which will be a shared parent block for event listing pages' `EventItemBlock` and service pages' `EventBlock`.
- Delete `event-item.{html|yaml}`, which are now unused.
- Add the following patterns to the pattern library to display variations:
    - `event-listing--empty.html` to show an empty event listing page, and
    - `listing--event--no-image.html` to show an event listing item with no image (to view the fallback image).

### Pending/Blockers

- [x] ~~Confirmed the copy for no upcoming/past events.~~ (Copy updated in [819e5ca](https://github.com/torchbox/torchbox.com/pull/190/commits/819e5ca7d52293c807d6070dca331bc6c653556c).)

### How to Test

- View the event listing page in the pattern library. ([local](http://localhost:8000/pattern-library/pattern/patterns/pages/events/events_listing.html))
- Create an event listing page, populate it and view the preview. ([local](http://localhost:8000/admin/pages/add/events/eventindexpage/5/))

### Screenshots

<details><summary>Mobile</summary>

image

</details>

<details><summary>Desktop</summary>

Unfortunately, I haven't figured out how to have pagination on the pattern library without sending it in as a separate context item.

image

</details>

<details><summary>With pagination</summary>

image

</details>

### MR Checklist

- [x] Add a description of your pull request and instructions for the reviewer to verify your work.
- [x] If your pull request is for a specific ticket, link to it in the description.
- [x] Stay on point and keep it small so the merge request can be easily reviewed.
- [x] Tests and linting passes.

#### Unit tests

- [x] Not required

#### Documentation

- [x] Updated editor guidelines (https://docs.google.com/document/d/1PAWccdQ4tfaZsrEWmpDhvP3GH5RRmBOARFVp4b-kje8/edit#heading=h.9bee8h7nj7kh)

#### Browser testing

- [x] I have tested in the following browsers and environments
  - Latest version of Chrome on mac
  - Latest version of Firefox on mac
  - Latest version of Safari on mac
  - Safari on last two versions of iOS (1 real, 1 on Browserstack)
  - Chrome on last two versions of Android (2 on Browserstack)

#### Data protection

- [x] Not relevant

#### Accessibility

- [x] Automated WCAG 2.1 tests pass
- [x] HTML validation passes
- [x] Manual WCAG 2.1 tests completed
- [x] I have tested in a screen reader
- [x] I have tested in high-contrast mode
- [x] ~~Any animations removed for prefers-reduced-motion~~

#### Sustainability

- [x] Images are optimised and lazy-loading used where appropriate
- [x] SVGs have been optimised
- [x] Perfomance and transfer of data considered
- [x] ~~If JavaScript is needed alternatives have been considered~~

#### Pattern library

- [x] The pattern library component for this template displays correctly, and does not break parent templates
- [x] The styleguide is updated if relevant

### CODE REVIEW

File: tbx/project_styleguide/templates/patterns/molecules/listing/listing--event--no-image.html [hunk 1]
```diff
@@ -0,0 +1,3 @@
+{# Component for preview only. #}
```
Comment (helenb) [inline tbx/project_styleguide/templates/patterns/molecules/listing/listing--event--no-image.html:1]: We have used the styleguide to display these kinds of variations everywhere else in the project, so I'd prefer not to add these pattern library only components here, and include them in the styleguide instead - see http://localhost:8000/pattern-library/render-pattern/patterns/styleguide/components/components.html

File: tbx/project_styleguide/templates/patterns/molecules/listing/listing--event.html [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+    {% image listing_image fill-740x600 format-webp saturation-0.6 as listing_mobile_image_retina %}
+{% endif %}
+
+<li class="listing-event listing-image listing {{ base_class }}">
+    {% if listing_image %}
+        {# alt text is empty for consistency with listing--avatar.html that includes avatar.html which has empty alt text #}
+        <picture class="listing-image__picture">
+            <source media="(max-width: 598px)" srcset="{{  listing_mobile_image.url }} 1x, {{ listing_mobile_image_retina.url }} 2x" />
+            <source media="(min-width: 599px)" srcset="{{ listing_desktop_image.url }} 1x, {{ listing_desktop_image_retina.url }} 2x" />
+            <img src="{{ listing_desktop_image.url }}" alt="" loading="lazy" class="listing-image__image" />
```
Comment (helenb) [inline tbx/project_styleguide/templates/patterns/molecules/listing/listing--event.html:22]: can we add the desaturation filter here too

File: tbx/static_src/sass/components/_listing-event.scss [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+        color: var(--color--heading);
+        font-weight: $weight--semibold;
+    }
+
+    &__details {
+        gap: $spacer-mini;
+    }
+
+    &__placeholder-image {
+        // Keep aspect-ratio up-to-date with the image sizes in the template.
```
Comment (helenb) [inline tbx/static_src/sass/components/_listing-event.scss:22]: Praise: good commenting

**Review by helenb — COMMENTED** [review]
Thanks @SharmaineLim - front-end looks good. I've just raised a couple of points that need addressing.

File: tbx/events/models.py [hunk 4]
```diff
@@ -21,8 +16,10 @@ class EventIndexPage(ColourThemeMixin, ContactMixin, SocialFields, Page):
     parent_page_types = ["torchbox.HomePage"]
     subpage_types = []
 
+    events = StreamField([("event", EventItemBlock())], blank=True, use_json_field=True)
```
Comment (engineervix) [inline tbx/events/models.py:19]: **Comment**: if, at the time of merging this to `main`, #187 would have already been merged, then [`use_json_field=True` could be omitted](https://docs.wagtail.org/en/latest/releases/6.0.html#streamfield-no-longer-requires-use-json-field-true).

File: tbx/events/models.py [hunk 5]
```diff
@@ -36,79 +33,62 @@ class EventIndexPage(ColourThemeMixin, ContactMixin, SocialFields, Page):
...
 
-    related_services = ParentalManyToManyField(
-        "taxonomy.Service", related_name="events"
-    )
+        try:
+            paged_events = paginator.page(page)
+        except PageNotAnInteger:
+            paged_events = paginator.page(1)
+        except EmptyPage:
+            paged_events = paginator.page(paginator.num_pages)
```
Comment (engineervix) [inline tbx/events/models.py:83]: **Suggestion**: Use the [paginator.get_page(page)](https://docs.djangoproject.com/en/dev/ref/paginator/#django.core.paginator.Paginator.get_page) method

Please see https://github.com/torchbox/torchbox.com/pull/185#discussion_r1537738309 for reference

**Review by engineervix — APPROVED** [review]
BE LGTM, just one suggestion

File: tbx/events/models.py (hunk 4 above)
Comment (SharmaineLim) [reply tbx/events/models.py:19]: I'll keep an eye out, thanks Victor.

Note to self: Keep thread unresolved until we're ready to merge the PR.

File: tbx/events/models.py (hunk 5 above)
Comment (SharmaineLim) [reply tbx/events/models.py:83]: TIL. Thanks!

File: tbx/project_styleguide/templates/patterns/molecules/listing/listing--event.html (hunk 2 above)
Comment (SharmaineLim) [reply tbx/project_styleguide/templates/patterns/molecules/listing/listing--event.html:22]: It should already be applied a few lines up:
https://github.com/torchbox/torchbox.com/blob/819e5ca7d52293c807d6070dca331bc6c653556c/tbx/project_styleguide/templates/patterns/molecules/listing/listing--event.html#L10-L13

Did you mean something else?

File: tbx/project_styleguide/templates/patterns/molecules/listing/listing--event--no-image.html (hunk 1 above)
Comment (SharmaineLim) [reply tbx/project_styleguide/templates/patterns/molecules/listing/listing--event--no-image.html]: I see! I didn't quite grok that these components & variants would go there too. Thanks! I've moved this into the components.html file in [cb47ff0](https://github.com/torchbox/torchbox.com/pull/190/commits/cb47ff0bf786d34caf18f350dbb2463f9132eb84).

<details><summary>Screenshot</summary>

image

</details>

File: tbx/events/models.py (hunk 5 above)
Comment (SharmaineLim) [reply tbx/events/models.py]: Updated in [763923d](https://github.com/torchbox/torchbox.com/pull/190/commits/763923db3645ba9aa63b74961aa3a27cecb6e5cf).

File: tbx/project_styleguide/templates/patterns/molecules/listing/listing--event.html (hunk 2 above)
Comment (SharmaineLim) [reply tbx/project_styleguide/templates/patterns/molecules/listing/listing--event.html:22]: Discussed on standup that I can resolve this.

File: tbx/events/models.py (hunk 4 above)
Comment (helenb) [reply tbx/events/models.py:19]: I'm going to go ahead and merge this - agreed with Victor that he'll update the wagtail 6 PR once this is merged.

**Pull request merged**